proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,324 @@
+  * https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/wal.md
+  * [Ganesh's blog](https://ganeshvernekar.com/blog/prometheus-tsdb-the-head-block/)
+
+> TL;DR: We need to break forward compatibility of the WAL format. This document proposes a data migration strategy for WAL and a lightweight ways we can be more transparent about those in each Prometheus release. 
+> 
+> We propose to version WAL by segments; with the version in the filename e.g.`00000002-v2`. For the migration, we propose the *LTS-enhanced [Two-Fold Migration Strategy](#how-two-fold-migration-strategy) with a `--storage.<...>.wal-version` flag for each Prometheus mode*.
+
+## Glossary
+
+For this document, I use the following wording:
+
+* **Prometheus Backward Compatibility**: When new Prometheus can read old WAL format.
+* **Prometheus Forward Compatibility**: When old Prometheus can read the new WAL format, which is required for the potential reverts.
+* **WAL Breaking Change**: Change to the WAL that will break Prometheus backward or forward compatibility, e.g. adding a new format that will replace the existing one.
+
+In this document, data "compatibility" means both:
+
+* Prometheus won't crash during compatible rollout or reverts. 
+* No data is lost during compatible rollouts or reverts.
+
+## Why
+
+Recently we discussed various improvements to Prometheus Write-Ahead-Log (WAL, also used as WBL for OOO) to:
+
+* Support new features (e.g. created timestamp (CT), native histograms with custom buckets (nhcb)).
+* Efficiency improvements and tech debt cleanup (e.g. parallelization/sharding, combining records, different decoding formats).
+
+The [current WAL data](https://github.com/prometheus/prometheus/blob/975d5d7357a220192fe1307b10ded9b35130ab1c/tsdb/docs/format/wal.md) is written into a handwritten variadic-encoding binary format. It's generally not versioned, and it does not support unknown fields or data except 3 specific places:
+
+* [record type](https://github.com/prometheus/prometheus/blob/e410a215fbe89b67f0e8edef9de25ede503ea4e0/tsdb/record/record.go#L38)
+* [metric type in metadata record](https://github.com/prometheus/prometheus/blob/e410a215fbe89b67f0e8edef9de25ede503ea4e0/tsdb/record/record.go#L111)
+* [metadata fields are arbitrary labels in metadata record](https://github.com/prometheus/prometheus/blob/e410a215fbe89b67f0e8edef9de25ede503ea4e0/tsdb/record/record.go#L608).
+
+Historically, we didn't hit major problems because we were only adding new semantic data (e.g. exemplars, metadata, new native histograms) as new records. 
+However, these days, we need to add features to existing data (e.g. custom buckets that will replace classic histograms, cleanup of histogram records or created timestamps to samples). Even if we create a new record for those and use it for new samples, any rollback will **lose that information as they appear unknown in the old version**.
```

Comment:
I don't see how we can do better than `Even if we create a new record for those and use it for new samples, any rollback will **lose that information as they appear unknown in the old version**`, even with the proposed changes, an old Prometheus version cannot do better than treating the new format data/records as unknown. as part of the forward compatibility.
So, in all cases, the old Prometheus version will lose that information as it doesn't know how to interpret it.